app/policies/redirect_policy.rb

```diff
@@ -0,0 +1,148 @@
+#
+# OpenProject is a fork of ChiliProject, which is a fork of Redmine. The copyright follows:
+# Copyright (C) 2006-2013 Jean-Philippe Lang
+# Copyright (C) 2010-2013 the ChiliProject Team
+#
+# This program is free software; you can redistribute it and/or
+# modify it under the terms of the GNU General Public License
+# as published by the Free Software Foundation; either version 2
+# of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+#
+# See doc/COPYRIGHT.rdoc for more details.
+#++
+
+require 'uri'
+require 'cgi'
+
+# This capsulates the validation of a requested redirect URL.
+#
+class RedirectPolicy
+  attr_reader :validated_redirect_url, :request
+
+  def initialize(requested_url, hostname:, default:, escape: true, return_escaped: true)
+    @current_host = hostname
+    @return_escaped = return_escaped
+
+    @requested_url = preprocess(requested_url, escape)
+    @default_url = default
+  end
+
+  ##
+  # Performs all validations for the requested URL
+  def valid?
+    return false if @requested_url.nil?
+
+    [
+      # back_url must not contain two consecutive dots
+      :no_upper_levels,
+      # Require the path to begin with a slash
+      :path_has_slash,
+      # do not redirect user to another host
+      :same_host,
+      # do not redirect user to the login or register page
+      :path_not_blacklisted,
+      # do not redirect to another subdirectory
+      :matches_relative_root
+    ].all? { |check| send(check) }
+  end
+
+  ##
+  # Return a valid redirect URI.
+  # If the validation check on the current back URL apply
+  def redirect_url
+    if valid?
+      postprocess(@requested_url)
+    else
+      @default_url
+    end
+  end
+
+  private
+
+  ##
+  # Preprocesses the requested redirect URL.
+  # - Escapes it when necessary
+  # - Tries to parse it
+  # - Escapes the redirect URL when requested so.
+  def preprocess(requested, escape)
+    url = escape ? URI.escape(CGI.unescape(requested.to_s)) : requested
+    URI.parse(url)
+  rescue URI::InvalidURIError => e
+    Rails.logger.warn("Encountered invalid redirect URL '#{requested}': #{e.message}")
+    nil
+  end
+
+  ##
+  # Postprocesses the validated URL
+  def postprocess(redirect_url)
+    # Remove basic auth credentials
+    redirect_url.userinfo = ''
+
+    if @return_escaped
+      redirect_url.to_s
+    else
+      URI.unescape(redirect_url.to_s)
+    end
+  end
+
+  ##
+  # Avoid paths with references to parent paths
+  def no_upper_levels
+    !@requested_url.path.include? '..'
```

Comment:
I actually had a brief look at the old code to check whether this was introduced by you or present before.
So good job on making the validation more obvious!

Making the URl absolute sounds like a good solution...